src/eigen.jl

```diff
@@ -221,188 +221,197 @@ end
-    p2 = abs2(a11 - q) + abs2(a22 - q) + abs2(a33 - q) + 2 * p1
-    p = sqrt(p2 / 6)
-    invp = inv(p)
-    b11 = (a11 - q) * invp
-    b22 = (a22 - q) * invp
-    b33 = (a33 - q) * invp
-    b12 = a12 * invp
-    b13 = a13 * invp
-    b23 = a23 * invp
-    B = SMatrix{3,3,S}((b11, conj(b12), conj(b13), b12, b22, conj(b23), b13, b23, b33))
-    r = real(det(B)) / 2
-
-    # In exact arithmetic for a symmetric matrix  -1 <= r <= 1
-    # but computation error can leave it slightly outside this range.
-    if (r <= -1)
-        phi = Sreal(pi) / 3
-    elseif (r >= 1)
-        phi = zero(Sreal)
-    else
-        phi = acos(r) / 3
-    end
-
-    eig3 = q + 2 * p * cos(phi)
-    eig1 = q + 2 * p * cos(phi + (2*Sreal(pi)/3))
-    eig2 = 3 * q - eig1 - eig3     # since tr(A) = eig1 + eig2 + eig3
-
-    if r > 0 # Helps with conditioning the eigenvector calculation
-        (eig1, eig3) = (eig3, eig1)
-    end
-
-    # Calculate the first eigenvector
-    # This should be orthogonal to these three rows of A - eig1*I
-    # Use all combinations of cross products and choose the "best" one
-    r₁ = SVector(a11 - eig1, a12, a13)
-    r₂ = SVector(conj(a12), a22 - eig1, a23)
-    r₃ = SVector(conj(a13), conj(a23), a33 - eig1)
-    n₁ = sum(abs2, r₁)
-    n₂ = sum(abs2, r₂)
-    n₃ = sum(abs2, r₃)
-
-    r₁₂ = r₁ × r₂
-    r₂₃ = r₂ × r₃
-    r₃₁ = r₃ × r₁
-    n₁₂ = sum(abs2, r₁₂)
-    n₂₃ = sum(abs2, r₂₃)
-    n₃₁ = sum(abs2, r₃₁)
-
-    # we want best angle so we put all norms on same footing
-    # (cheaper to multiply by third nᵢ rather than divide by the two involved)
-    if n₁₂ * n₃ > n₂₃ * n₁
-        if n₁₂ * n₃ > n₃₁ * n₂
-            eigvec1 = r₁₂ / sqrt(n₁₂)
-        else
-            eigvec1 = r₃₁ / sqrt(n₃₁)
-        end
-    else
-        if n₂₃ * n₁ > n₃₁ * n₂
-            eigvec1 = r₂₃ / sqrt(n₂₃)
+    function get_cos_sin(u::T,v::T) where {T}
+        max_abs = max(abs(u), abs(v))
+        if max_abs > 0
+            u,v = (u,v) ./ max_abs
+            len = sqrt(u^2 + v^2)
+            cs, sn = (u,v) ./ len
+            if cs > 0
+                cs = -cs
+                sn = -sn
+            end
+            T(cs), T(sn)
         else
-            eigvec1 = r₃₁ / sqrt(n₃₁)
+            T(-1), T(0)
         end
     end
 
-    # Calculate the second eigenvector
-    # This should be orthogonal to the previous eigenvector and the three
-    # rows of A - eig2*I. However, we need to "solve" the remaining 2x2 subspace
-    # problem in case the cross products are identically or nearly zero
-
-    # The remaing 2x2 subspace is:
-    @inbounds if abs(eigvec1[1]) < abs(eigvec1[2]) # safe to set one component to zero, depending on this
-        orthogonal1 = SVector(-eigvec1[3], zero(S), eigvec1[1]) / sqrt(abs2(eigvec1[1]) + abs2(eigvec1[3]))
-    else
-        orthogonal1 = SVector(zero(S), eigvec1[3], -eigvec1[2]) / sqrt(abs2(eigvec1[2]) + abs2(eigvec1[3]))
+    function _sortperm3(v)
+        local perm = SVector(1,2,3)
```

Comment:
This seems ok as a workaround... though do note that having inner functions inside a complex implementation like `_eig` without the intent to have them capture any state always runs the risk that you will accidentally share bindings. For example, if `max_abs` in `get_cos_sin` happened to be a variable name in `_eig` you would again hit the same problem.

If you don't intend to use these as closures (ie, you don't want to capture any state from the outer function) it's a fragile solution to add `local` on a piecemeal basis, overall I think it would be better to just move these helper functions outside of `_eig`. Put another way: the placement of `_sortperm3` is *quite* significant in ways other than its visibility outside of `_eig`: I believe the closure-or-not distinction should be considered as a priority vs the other side effect of placement — whether `_sortperm3` is visible at the module level.

Reply:
(I do realize I suggested adding `local` as one of the options to fix this :grimacing:. But on further consideration I think I was wrong about that. Unless I've missed something and you really do capture some state here.)

Reply:
Yeah, it's a thing I wasn't too happy about either... was more of a "getting this fixed and testable as fast as possible" type of deal.

Would be cool to have some kind of way to annotate an inner function as "non-capturing" / "only here because it isn't used elsewhere".

Reply:
I agree with Chris's general advice here (it's a shame scoping can become a footgun...) - but I wouldn't block this PR on that. Are there any other outstanding issues?

Reply:
Yeah, it's not blocked on that; more just that I'm loosing track of things.